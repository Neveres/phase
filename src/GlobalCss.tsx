import React from 'react'
import { css, Global } from '@emotion/react'

export const GlobalCss = () => (
  <Global
    styles={css`
      body {
        overflow: hidden;
      }
    `}
  />
)

export const commentsContainer = css`
  .comment-header {
    display: flex;
    justify-content: space-between;

    font-size: 16px;

    .comment-name {
      font-weight: 700;
    }
  }

  .comment-message {
    margin-bottom: 5px;
  }

  margin-bottom: 15px;
  border-bottom: 1px solid #f0f0f0;
`
