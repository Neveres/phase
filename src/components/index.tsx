export { ReactPixiStage } from './ReactPixiStage'
export { CommentEntries } from './CommentEntries'
export { CommentDialog } from './CommentDialog'
export { AppHeader } from './AppHeader'
export { LoginDialog } from './LoginDialog'
