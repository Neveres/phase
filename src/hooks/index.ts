export * from './useCommentGroups'
export * from './useStorage'
