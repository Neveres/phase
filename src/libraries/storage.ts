export const getItem = (
  key: Parameters<typeof localStorage.getItem>[0],
  parse = false,
) => {
  const value = localStorage.getItem(key) || ''
  if (parse) {
    let parsedValue
    try {
      parsedValue = JSON.parse(value)
    } catch (ignore) {
      parsedValue = value
    }
    return parsedValue
  } else {
    return value
  }
}

export const setItem = (
  key: string,
  value: Parameters<typeof JSON.stringify>[0],
) => {
  localStorage.setItem(
    key,
    typeof value !== 'string' ? JSON.stringify(value) : value,
  )
}

export const removeItem = (
  key: Parameters<typeof localStorage.removeItem>[0],
) => {
  return localStorage.removeItem(key)
}

export const clear = () => {
  localStorage.clear()
}
