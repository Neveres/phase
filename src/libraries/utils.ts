export const generateUuid = () => {
  let nowOfDate = Date.now()

  if (
    typeof performance !== 'undefined' &&
    typeof performance.now === 'function'
  ) {
    nowOfDate += performance.now()
  }

  const result = 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(
    /[xy]/g,
    (character) => {
      const remains = (nowOfDate + Math.random() * 16) % 16 | 0
      nowOfDate = Math.floor(nowOfDate / 16)
      return (character === 'x' ? remains : (remains & 0x3) | 0x8).toString(16)
    },
  )
  return result
}
