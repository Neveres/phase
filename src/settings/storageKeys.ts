export const storageKeys = {
  COMMENT_GROUPS: 'COMMENT_GROUPS',
  USERNAME: 'USERNAME',
}
