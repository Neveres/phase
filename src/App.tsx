import React, { useState, useCallback } from 'react'
import { Modal, Input } from 'antd'
import { ReactPixiStage } from 'src/components'
import { useComment } from 'src/hooks'

const USER_NAME = 'user'
const App = () => {
  const [coordinate, setCoordinate] = useState([] as number[])
  const [isModalOpen, setIsModalOpen] = useState(false)
  const [message, setMessage] = useState('')
  const { comments, addComment } = useComment()

  const openModal = useCallback(() => {
    setIsModalOpen(true)
  }, [])

  const closeModal = useCallback(() => {
    setIsModalOpen(false)
  }, [])

  const openModalAndSetCoordinate = useCallback(
    (newCoordinate: number[]) => {
      openModal()
      setCoordinate(newCoordinate)
    },
    [openModal],
  )

  const clearMessage = useCallback(() => {
    setMessage('')
  }, [])

  const onOk = useCallback(() => {
    closeModal()

    if (message) {
      addComment({
        name: USER_NAME,
        message,
        coordinate,
      })
      clearMessage()
    }
  }, [addComment, clearMessage, closeModal, coordinate, message])

  const onCancel = useCallback(() => {
    closeModal()
    clearMessage()
  }, [clearMessage, closeModal])

  const onChangeMessage = useCallback(
    ({ target: { value } }: React.ChangeEvent<HTMLInputElement>) => {
      setMessage(value)
    },
    [],
  )

  return (
    <>
      <ReactPixiStage setCoordinate={openModalAndSetCoordinate} />
      <Modal
        title="Add comments"
        open={isModalOpen}
        onOk={onOk}
        onCancel={onCancel}
      >
        <Input value={message} onChange={onChangeMessage} />
      </Modal>
    </>
  )
}

export default App
