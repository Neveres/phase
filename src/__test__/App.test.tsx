import React from 'react'
import { create } from 'react-test-renderer'
import { render, screen, fireEvent, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { setItem, clear } from 'src/libraries'
import { storageKeys } from 'src/settings'
import App from '../App'

afterAll(() => {
  clear()
})

describe('App', () => {
  test('should render well', () => {
    expect(create(<App />).toJSON()).toMatchSnapshot()
  })

  test('Opening login dialog should work well', async () => {
    setItem(storageKeys.USERNAME, 'username')
    render(<App />)
    fireEvent.click(screen.getByText('Logout'))

    await waitFor(() => {
      expect(screen.getByTitle('Username')).toBeInTheDocument()
    })

    fireEvent.change(screen.getByTestId('username-input'), {
      target: { value: 'username' },
    })

    fireEvent.change(screen.getByTestId('password-input'), {
      target: { value: 'password' },
    })

    fireEvent.click(screen.getByText('Log In'))

    await waitFor(() => {
      expect(screen.queryByText('Username')).not.toBeInTheDocument()
    })
  })

  describe('Opening comment dialog should work well', () => {
    test('while adding more comment', async () => {
      setItem(storageKeys.USERNAME, 'username')
      setItem(storageKeys.COMMENT_GROUPS, {
        uuid: {
          uuid: 'uuid',
          comments: [
            { name: 'name', message: 'message', postTime: 'postTime' },
          ],
          isResolved: false,
          coordinate: [100, 100],
        },
      })
      render(<App />)

      fireEvent.click(screen.getByRole('img'))

      await waitFor(() => {
        expect(screen.getByText('OK')).toBeInTheDocument()
      })

      fireEvent.click(screen.getByText('Cancel'))

      await waitFor(() => {
        expect(screen.queryByText('OK')).not.toBeInTheDocument()
      })
    })

    test('while create new comment', async () => {
      setItem(storageKeys.USERNAME, 'username')
      render(<App />)

      userEvent.click(document.querySelector('canvas') as any)
      await waitFor(() => {
        expect(screen.getByText('OK')).toBeInTheDocument()
      })
    })
  })
})
