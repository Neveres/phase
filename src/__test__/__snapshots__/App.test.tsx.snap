// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render well 1`] = `
Array [
  <div
    className="css-1afp8zc-CommentEntries"
  />,
  <canvas
    onClick={[Function]}
  />,
]
`;
